pkg/controller/migplan/validation.go

```diff
@@ -595,6 +611,135 @@ func (r ReconcileMigPlan) validateConflict(plan *migapi.MigPlan) error {
 	return nil
 }
 
+// Validiate Namespace Selection
+func (r ReconcileMigPlan) validateNamespaceAuthorization(plan *migapi.MigPlan) error {
+	unauthorizedNamespaces := make([]string, 0)
+
+	if plan.Status.HasAnyCondition(Suspended) {
+		return nil
+	}
+
+	cluster, err := plan.GetSourceCluster(r)
+	if err != nil {
+		log.Trace(err)
+		return err
+	}
+
+	if cluster == nil || !cluster.Status.IsReady() {
+		return nil
+	}
+
+	err = r.ValidateIdentities(plan)
+	if err != nil {
+		plan.Status.SetCondition(migapi.Condition{
+			Type:     IdentitySecretMisconfigured,
+			Status:   True,
+			Reason:   KeyNotFound,
+			Category: Error,
+			Message:  IdentitySecretMisconfiguredMessage,
+		})
+		log.Trace(err)
+		return err
+	}
+
+	err = r.SetIdentities(plan)
+	if err != nil {
+		log.Trace(err)
+		return err
+	}
+
+	authorized, err := r.SrcIdentity.HasRead(plan.Spec.Namespaces)
+	if err != nil {
+		// This error comes when the given source identity token failed to
+		// authenticate with the source cluster
+		plan.Status.SetCondition(migapi.Condition{
+			Type:     SourceIdentityTokenNotAuthorized,
+			Status:   True,
+			Reason:   NotAuthorized,
+			Category: Error,
+			Message:  SourceIdentityTokenNotAuthorizedMessage,
+		})
+		log.Trace(err)
+		return err
+	}
+
+	for ns, auth := range authorized {
+		if !auth {
+			unauthorizedNamespaces = append(unauthorizedNamespaces, ns)
+		}
+	}
+
+	if len(unauthorizedNamespaces) > 0 {
+		plan.Status.SetCondition(migapi.Condition{
+			Type:     InvalidNamespaceAuthorization,
+			Status:   True,
+			Reason:   NotAuthorized,
+			Category: Error,
+			Message:  InvalidNamespaceAuthorizationMessage,
+			Items:    unauthorizedNamespaces,
+		})
+		return errors.New("selected namespaces are not authorized")
+	}
+
+	return nil
+}
+
+// Validate Identity Tokens on Identity struct
+// Returns an error if secret is somehow misconfigured
+func (r *ReconcileMigPlan) ValidateIdentities(plan *migapi.MigPlan) error {
+	_, err := plan.GetSourceIdentity(r.Client)
+	if err != nil {
+		return err
+	}
+	_, err = plan.GetDestinationIdentity(r.Client)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Sets identities on ReconcileMigPlan struct
+// Returns if already set
+func (r *ReconcileMigPlan) SetIdentities(plan *migapi.MigPlan) error {
```

Comment:
Seems unexpected for this to be done here (in validation).

Reply:
Good point, I will move this to `Reconcile`